lib/reconstruct.ts

```diff
@@ -0,0 +1,73 @@
+import { types, Root, Token, Tokens, Group } from './types';
+import * as sets from './sets';
+
+const simplifications: [sets.SetFunc, string][] = [
+  [sets.words, '\\w'],
+  [sets.notWords, '\\W'],
+  [sets.ints, '\\d'],
+  [sets.notInts, '\\D'],
+  [sets.whitespace, '\\s'],
+  [sets.notWhitespace, '\\S'],
+  [sets.anyChar, '.']
+];
+
+const reduceStack = (stack: Token[]): string => stack.map(reconstruct).join('');
+
+const createAlternate = (token: Root | Group): string => {
+  if ('options' in token) {
+    return token.options.map(reduceStack).join('|');
+  } else if ('stack' in token) {
+    return reduceStack(token.stack);
+  } else {
+    throw new Error(`options or stack must be Root or Group token`);
+  }
+};
+
+export const reconstruct = (token: Tokens): string => {
+  switch (token.type) {
+    case types.ROOT:
+      return createAlternate(token);
+    case types.CHAR:
+      const c = String.fromCharCode(token.value);
+      return (/[[\]^.\/|?*+()]/.test(c) ? '\\' : '') + c;
+    case types.POSITION:
+      if (token.value === '^' || token.value === '$') {
+        return token.value;
+      } else {
+        return '\\' + token.value;
+      }
+    case types.REFERENCE:
+      return '\\' + token.value;
+    case types.SET:
+      for (const [set, simplification] of simplifications) {
+        if (JSON.stringify(set()) === JSON.stringify(token)) {
+          return simplification;
+        }
+      }
+      return `[${token.not ? '^' : ''}${reduceStack(token.set)}]`;
+    case types.RANGE:
+      return `${String.fromCharCode(token.from)}-${String.fromCharCode(token.to)}`;
+    case types.GROUP:
+      // Check token.remember
+      return `(${token.remember ? '' : '?'}${token.followedBy ? '=' :
+          token.notFollowedBy ? '!' :
+            (token.remember ? '' : ':')
+        }${createAlternate(token)})`
```

Comment:
Potentially more readable without trying to share the `?`.

```suggestion
      const prefix =
        token.remember ? '' :
        token.followedBy ? '?=' :
        token.notFollowedBy ? '?!' :
        '?:';

      return `(${prefix}${createAlternate(token)})`;
```